cli/internal/upgrade/iammigrate.go

```diff
@@ -0,0 +1,112 @@
+/*
+Copyright (c) Edgeless Systems GmbH
+
+SPDX-License-Identifier: AGPL-3.0-only
+*/
+
+package upgrade
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"path/filepath"
+	"strings"
+
+	"github.com/edgelesssys/constellation/v2/cli/internal/terraform"
+	"github.com/edgelesssys/constellation/v2/internal/cloud/cloudprovider"
+	"github.com/edgelesssys/constellation/v2/internal/constants"
+	"github.com/edgelesssys/constellation/v2/internal/file"
+)
+
+// TfMigrationCmd is an interface for all terraform upgrade / migration commands.
+type TfMigrationCmd interface {
+	CheckTerraformMigrations(file file.Handler) error
+	Plan(ctx context.Context, file file.Handler, outWriter io.Writer) (bool, error)
+	Apply(ctx context.Context, fileHandler file.Handler) error
+	String() string
+	UpgradeID() string
+}
+
+// IAMMigrateCmd is a terraform migration command for IAM.
+type IAMMigrateCmd struct {
+	tf        tfIAMClient
+	upgradeID string
+	csp       cloudprovider.Provider
+	logLevel  terraform.LogLevel
+}
+
+// NewIAMMigrateCmd creates a new IAMMigrateCmd.
+func NewIAMMigrateCmd(ctx context.Context, upgradeID string, csp cloudprovider.Provider, logLevel terraform.LogLevel) (*IAMMigrateCmd, error) {
```

Comment:
Regarding https://github.com/edgelesssys/constellation/pull/2090#discussion_r1267848167 and https://github.com/edgelesssys/constellation/pull/2090#discussion_r1266949703.
If we expect `upgradeID string` here we loose all gurantees of creating a variable of type `upgradeID` in the first place. Exporing it and expecting `kubernetes.UpgradeID` here creates a very likely unwanted dependency. Since we always call `kubernetes.NewUpgradeID(...)` directly before calling `kubernetes.NewUpgrader(...)` maybe we can create the upgrateID directly in `kubernetes.NewUpgrader(...)` and then have a getter, have the field be public, or simply return it as a string from the constructor. This has the same gurantees as the current solution while completely removing the need of enforing a type.